src/nix-env/nix-env.cc

```diff
@@ -1368,7 +1369,9 @@ static int main_nix_env(int argc, char * * argv)
         Globals globals;
 
         globals.instSource.type = srcUnknown;
-        globals.instSource.nixExprPath = getHome() + "/.nix-defexpr";
+        Path nixExprPath = getHome() + "/.nix-defexpr";
+        if (!pathExists(nixExprPath)) nixExprPath = getStateDir() + "/nix/defexpr";
+        globals.instSource.nixExprPath = nixExprPath;
```

Comment:
I like to keep temporaries narrowly scoped. Easier to read later because you can hold less in head.
```suggestion
        {
            Path nixExprPath = getHome() + "/.nix-defexpr";
            if (!pathExists(nixExprPath)) nixExprPath = getStateDir() + "/nix/defexpr";
            globals.instSource.nixExprPath = nixExprPath;
        }
```